hkmc2/shared/src/main/scala/hkmc2/semantics/ImplicitResolver.scala

```diff
@@ -0,0 +1,108 @@
+package hkmc2.semantics
+
+import mlscript.utils.*, shorthands.*
+import hkmc2.utils.TraceLogger
+import hkmc2.Raise
+import hkmc2.semantics.Elaborator.State
+import hkmc2.syntax.Ins
+import hkmc2.ErrorReport
+import hkmc2.Message.MessageContext
+import hkmc2.syntax.Tree.TermDef
+import hkmc2.syntax.Fun
+
+case class CtxArgImpl(private val param: Param) extends CtxArg:
```

Comment:
This should probably not be a case class because it's basically a state wrapper, not a plain data holder.